src/main/fc/fc_msp.c

```diff
@@ -182,14 +183,65 @@ typedef enum {
 #define RATEPROFILE_MASK (1 << 7)
 
 #ifdef USE_SERIAL_4WAY_BLHELI_INTERFACE
+#define ESC_4WAY 0xff
+
+uint8_t escMode;
+uint8_t escPortIndex = 0;
+
 static void msp4WayIfFn(serialPort_t *serialPort)
 {
-    // rem: App: Wait at least appx. 500 ms for BLHeli to jump into
-    // bootloader mode before try to connect any ESC
-    // Start to activate here
-    esc4wayProcess(serialPort);
-    // former used MSP uart is still active
-    // proceed as usual with MSP commands
+#ifdef USE_ESCSERIAL
+    if (escMode != ESC_4WAY) {
+        escEnablePassthrough(serialPort, escPortIndex, escMode);
+    } else {
+#else
+    {
+#endif
+        // rem: App: Wait at least appx. 500 ms for BLHeli to jump into
+        // bootloader mode before try to connect any ESC
+        // Start to activate here
+        esc4wayProcess(serialPort);
+        // former used MSP uart is still active
+        // proceed as usual with MSP commands
+    }
+}
+
+static void mspFc4waySerialCommand(sbuf_t *dst, sbuf_t *src, mspPostProcessFnPtr *mspPostProcessFn)
+{
+    const unsigned int dataSize = sbufBytesRemaining(src);
+    if (dataSize == 0) {
+        // Legacy format
+
+        escMode = ESC_4WAY;
+    } else {
+        escMode = sbufReadU8(src);
+        escPortIndex = sbufReadU8(src);
+    }
+
+    if (escMode == ESC_4WAY) {
+        // get channel number
+        // switch all motor lines HI
+        // reply with the count of ESC found
+        sbufWriteU8(dst, esc4wayInit());
+
+        if (mspPostProcessFn) {
+            *mspPostProcessFn = msp4WayIfFn;
+        }
+    } else {
+#ifdef USE_ESCSERIAL
```

Comment:
It's better to (at least) style the preprocessor conditionals so that C block are balanced - your code has more `{` than `}` confusing editors (and making ccode hard to read).